### DISCUSSION TITLE
Injecting data into getServerSideProps?

### DISCUSSION THREAD

**Discussion by ryanbahan** [vercel/next.js#14775] [Help]
I'm wondering if it's possible to inject additional data into getServerSideProps? I'm basically looking to have access to client-side data, as well as data being held in a Context API store, when switching between pages (like you can do with Redux; accessing the store on the server).

I know there are request headers for cookies that can partially solve this data-passing, but moreso I'm wondering if there's a way to take pieces of local data and "pass" it to the getServerSideProps method so that I can use that data for other fetches, etc.

**Comment by pkrawc** [comment]
Can you give an example of what kind of data you're wanting to pass into `getServerSideProps`? Cookies would suffice for every use case I can think of.

**Reply by ryanbahan** [reply]
I basically have a large payload of data that I store in Context API on initial page load of my app. When switching to a new pages, there are scenarios where I'd rather not refetch data I already have, and I also want to pre-render that data server-side instead of using useContext hooks. 

The problem in that I can't access Context API in getServerSideProps, so I'm trying to think through how I can pass that context data between pages and implement SSR. Basically looking for the equivalent of the Redux wrapper for Nextjs, or some way to inject the data server-side.

**Reply by pkrawc** [reply]
You could just load the data in a custom _app.js page. You can probably just load the data in getServerSideProps and populate the initial rendering of your context with that data.

If you're just doing client-side page transitions, the _app.js component should not remount, and that would allow you to just feed the data into all your pages. If you're doing a hard refresh then you're going to need to load the data again because the application state is lost anyway.

**Reply by ryanbahan** [reply]
Maybe I'm wrong, but I thought that using getServerSideProps in _app was discouraged as it disables automatic static optimization?

**Reply by ryanbahan** [reply]
This seems like a reasonable solution, though another consideration is that I'm actively manipulating that data on the client side after it loads - so I'm not sure this use case would be ideal for that. Having the data in a store where I can access it on both client + server as opposed to loading initially would definitely be the ideal scenario. I may just have to transition to Redux to accomplish that, I guess.

**Comment by andreisena** [comment]
What about a HOF on top of getServerSideProps?

```js
const withData = fn => ctx => {
  // Fetch data

  return fn({ ...ctx, data })
}
```

This way you could use on specific pages without breaking SSG on all pages:

```js
const getServerSideProps = withData(({ data }) => {

  // ...

  return {
    props: {
      data,
    }
  }
})
```

**Reply by NekfarBaqir** [reply]
As you know that getServerSideProps is an async function, mine gave me error : getServerSideProps isn't a function.

**Comment by RossMcMillan92** [comment]
This was driving me mental. Think I got a workaround: https://github.com/vercel/next.js/discussions/10874#discussioncomment-3091438
